internal/db/db.go

```diff
@@ -143,115 +144,253 @@ func (tx *holdTx) FindOne(modelID uuid.UUID, matcher Matcher) (rec Record, err e
 	return
 }
 
-func (tx holdTx) FindMany(modelID uuid.UUID, matcher Matcher) []Record {
-	mi := tx.h.IterMatches(modelID, matcher)
-	hits := []Record{}
-	for val, ok := mi.Next(); ok; val, ok = mi.Next() {
-		hits = append(hits, val)
+func (tx *holdTx) FindMany(modelID uuid.UUID, matcher Matcher) (recs []Record, err error) {
+	recs, err = tx.h.FindMany(modelID, matcher)
+	return
+}
+
+func (tx *holdTx) Insert(rec Record) error {
+	tx.ensureWrite()
+	h, err := tx.h.Insert(rec)
+	if err != nil {
+		return err
 	}
-	return hits
+	tx.h = h
+	return nil
 }
 
-func (tx *holdTx) Insert(rec Record) {
+func (tx *holdTx) Update(r Record, fields map[string]interface{}) (Record, error) {
 	tx.ensureWrite()
-	tx.h = tx.h.Insert(rec)
+	//Need to load a model for the record
+	m, _ := tx.GetModel(r.Type())
+	newRec := r.DeepCopy(m)
+	for key, val := range fields {
+		err := newRec.Set(key, val)
+		if err != nil {
+			return nil, err
+		}
+	}
+	h, err := tx.h.Insert(newRec)
+	if err != nil {
+		return nil, err
+	}
+	tx.h = h
+	return newRec, nil
 }
 
-func (tx *holdTx) Connect(left, right Record, rel Relationship) {
+func (tx *holdTx) Connect(left, right Record, rel Relationship) error {
 	tx.ensureWrite()
 
 	if rel.LeftBinding == BelongsTo && (rel.RightBinding == HasOne || rel.RightBinding == HasMany) {
 		// FK left
-		left.SetFK(rel.LeftName, right.ID())
+		id, err := right.ID()
+		if err != nil {
+			return err
+		}
+		err = left.SetFK(rel.LeftName, id)
+		if err != nil {
+			return err
+		}
 	} else if rel.RightBinding == BelongsTo && (rel.LeftBinding == HasOne || rel.LeftBinding == HasMany) {
 		// FK right
-		right.SetFK(rel.RightName, left.ID())
+		id, err := left.ID()
+		if err != nil {
+			return err
+		}
+		right.SetFK(rel.RightName, id)
+		if err != nil {
+			return nil
+		}
 	} else if rel.LeftBinding == HasManyAndBelongsToMany && rel.RightBinding == HasManyAndBelongsToMany {
 		// Join table
 		panic("Many to many relationships not implemented yet")
 	} else {
-		panic("Trying to connect invalid relationship")
+		return fmt.Errorf("Trying to connect invalid relationship")
+	}
+	h1, err := tx.h.Insert(left)
+	if err != nil {
+		return err
+	}
+	h2, err := h1.Insert(right)
+	if err != nil {
+		return err
 	}
-	h1 := tx.h.Insert(left)
-	h2 := h1.Insert(right)
 	tx.h = h2
+	return nil
 }
 
-func LoadRel(storeRel Record) Relationship {
-	return Relationship{
-		ID:           storeRel.ID(),
-		LeftBinding:  RelType(storeRel.Get("leftBinding").(int64)),
-		LeftModelID:  storeRel.GetFK("leftModel"),
-		LeftName:     storeRel.Get("leftName").(string),
-		RightBinding: RelType(storeRel.Get("rightBinding").(int64)),
-		RightModelID: storeRel.GetFK("rightModel"),
-		RightName:    storeRel.Get("rightName").(string),
+func LoadRel(storeRel Record) (Relationship, error) {
+	id, err := storeRel.ID()
```

Comment:
boy, this sure is noisy. But I think you can clean it up pretty nicely by following the `errWriter` pattern in this golang.org blog post: https://blog.golang.org/errors-are-values

(RecordReader or something like that)